merlin/measurement/photon_loss.py

```diff
@@ -0,0 +1,348 @@
+        loss_keys: list[tuple[int, ...]] = []
+        row_entries: list[dict[int, float]] = []
+
+        for sim_key in self._simulation_keys:
+            per_mode = [
+                self._mode_loss_distribution(count, self._survival_probs[mode])
+                for mode, count in enumerate(sim_key)
+            ]
+
+            combined: dict[int, float] = {}
+            for outcomes in itertools.product(*per_mode):
+                loss_key = tuple(outcome[0] for outcome in outcomes)
+                probability = 1.0
+                for _, partial_prob in outcomes:
+                    probability *= partial_prob
+
+                if probability == 0.0:
+                    continue
+
+                column_index = key_to_index.get(loss_key)
+                if column_index is None:
+                    column_index = len(loss_keys)
+                    key_to_index[loss_key] = column_index
+                    loss_keys.append(loss_key)
+                combined[column_index] = combined.get(column_index, 0.0) + probability
+
+            row_entries.append(combined)
+
+        rows = len(self._simulation_keys)
+        cols = len(loss_keys)
+        device_kwargs = {}
+        if self._device is not None:
+            device_kwargs["device"] = self._device
+
+        matrix = torch.zeros(
+            (rows, cols),
+            dtype=self._dtype,
+            device=device_kwargs.get("device", None),
+        )
+        for row_idx, entries in enumerate(row_entries):
+            for col_idx, prob in entries.items():
+                matrix[row_idx, col_idx] = prob
+
+        return matrix, loss_keys, False
+
+    # ------------------------------------------------------------------
+    # Public API
+    # ------------------------------------------------------------------
+
+    @property
+    def output_keys(self) -> list[tuple[int, ...]]:
+        """Classical Fock keys after photon loss."""
+        return self._loss_keys
+
+    @property
+    def output_size(self) -> int:
+        """Number of classical outcomes after photon loss."""
+        return len(self._loss_keys)
+
+    @property
+    def is_identity(self) -> bool:
+        """Whether the transform corresponds to perfect transmission."""
+        return self._is_identity
+
+    def forward(self, distribution: torch.Tensor) -> torch.Tensor:
+        """
+        Apply the photon loss transform to a Fock probability vector.
+
+        Args:
+            distribution: A Fock probability vector as a 1D torch tensor.
+
+        Returns:
+            A Fock probability vector after photon loss.
+        """
+        if self._is_identity:
+            return distribution
+
+        matrix: torch.Tensor = cast(torch.Tensor, self._matrix)  # type: ignore[attr-defined]
+        if distribution.dtype != matrix.dtype:
+            matrix = matrix.to(distribution.dtype)
+        if distribution.device != matrix.device:
+            matrix = matrix.to(distribution.device)
```

Comment:
global note here - for a later review, normally nn module do not move/cast  their internal data to what is needed for processing - if they have not been moved before - they simply fail. So my intuition is that we should not do any of this "smart" moving. what do you think?